### PR TITLE
[Adjustments] TaxRate changes

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Alright, this is the big one. If backporting the adjustments improvements from Spree 2.2 were a computer game, this would be the final boss. This PR is basically two parts (in terms of the commits) but they both need to be done at the same time.

**First part**

Brings in substantial changes to `Spree::TaxRate` from the `2-2-stable` branch. There are some updates to the interface here that mean that instead of passing an order to `TaxRate#adjust`, we now pass _items_. This means we can now apply (or recalculate) taxes on sets of items, or on a single item at a time instead of everything all at once (moving towards more granularity). The main place this is called from is `Order#create_tax_charge!`.

**Second part**

Moves the responsibility for taxing **enterprise fees** from a method buried in `EnterpriseFeeApplicator` to being handled by `Spree::TaxRate#adjust` in the same way as other taxes. Also means that taxes on enterprise fees are now recorded nicely in proper tax adjustments.

**Additional and Included taxes**

This PR also closes the huge gap between the way taxes are recorded for instances that use "included" taxes (EU and AUS) and instances that use "additional" taxes (US and CA). There's some more background on that in [this discourse post](https://community.openfoodnetwork.org/t/adjustments-changes/2159/7)

The TLDR version is: "included" tax instances record individual tax amounts per-item (on line items and fees, etc) which allows a high granularity and clarity when it comes to things like reports, but "additional" tax instances don't record granular tax amounts on items at all, but instead record a single sum per tax rate and put it on the order object. If you look through the tax data from both types of instance you see what looks like an entirely different data model :see_no_evil:

So after this PR, _all_ instances will have the same "granular" tax amounts recorded per-item. EU and AUS do this already, but it's a big change for US and CA. If you read through the aforementioned [discourse post](https://community.openfoodnetwork.org/t/adjustments-changes/2159/7), this is "Option 1". This means that orders created before this PR is deployed will still have lump-sum tax adjustments per order, and orders created afterwards will have nice granular tax amounts for all items. Our current reports and places where tax adjustments are displayed can handle both types. If an order with "legacy" additional taxes is "re-calculated" (via the admin order edit page) the "old style" taxes will be removed and replaced with the "new style" taxes. This option avoids doing a huge data migration to convert all the "old" taxes (Option 2), which looks like it would be really problematic.

#### What should we test?

This will require a lot of testing on our two new "included tax" and "additional tax" staging servers (FR and UK).

Testing plan here: https://github.com/openfoodfoundation/openfoodnetwork/issues/7429#issuecomment-854979259

#### Release notes

Updated TaxRate and taxing of enterprise fees. Instances with included tax will now have tax amounts recorded per-item.

Changelog Category: User facing changes

### CODE REVIEW

File: app/controllers/spree/admin/orders_controller.rb
Comment (sauloperez) [inline app/controllers/spree/admin/orders_controller.rb:43]: it may be too much to ask to extract a service to avoid adding yet more responsibilities to the controller, but this is growing and we should do it soon IMO.

File: engines/order_management/app/services/order_management/order/updater.rb
Comment (sauloperez) [inline engines/order_management/app/services/order_management/order/updater.rb:62]: finally! :heart_eyes:

File: app/services/order_fees_handler.rb
Comment (sauloperez) [inline app/services/order_fees_handler.rb:22]: I'm a bit concerned about the potential lock contention caused by this growing transaction. Is there a chance we can move the calls to `order.updater` below outside the block? It looks like a sensible trade-off to me.

File: app/models/spree/adjustment.rb
Comment (sauloperez) [inline app/models/spree/adjustment.rb:38]: I'm not following. Why would an adjustment have adjustments :thinking: ?

**Comment by sauloperez** [comment]
So "Option 1" means will have to support the two separate implementations, with the code complexity that entails, forever? even if we go with this option, we need to sunset additional taxes at some point.

File: app/models/spree/order.rb
Comment (sauloperez) [inline app/models/spree/order.rb:377]: IMO this might be the right time to extract `#create_tax_charge!` out of the `after_create` callback. Although some `before_*` callbacks might be legit, I tend to find the `after_*` rather code-smelly. This has been discussed in the Rails community in-depth in places like [Ruby science](https://github.com/thoughtbot/ruby-science/blob/47ae712a002c64e0d6ede3e4eb98fcfeaa9a4d39/book/code_smells/callback.md) or [here](https://www.thoughtbot.com/upcase/videos/callbacks).

Moving it to a service and calling it from `OrderWorkflow#after_transition_hook` would be a good starting point IMO. We should call it when `order.state == "payment"` if I'm not mistaken.

File: app/models/spree/order.rb
Comment (Matt-Yorkley) [reply app/models/spree/order.rb:70]: Good shout :+1: I thought it was going to be called from other places when I first added it.

File: engines/order_management/app/services/order_management/order/updater.rb
Comment (Matt-Yorkley) [reply engines/order_management/app/services/order_management/order/updater.rb:62]: :fire: :smile:

File: app/services/order_fees_handler.rb
Comment (Matt-Yorkley) [reply app/services/order_fees_handler.rb:22]: Yeah, I'll take a look at it. I'm not 100% convinced the lock is even useful here...?

File: app/models/spree/order.rb
Comment (Matt-Yorkley) [reply app/models/spree/order.rb:377]: I've been thinking this as well. I'm not sure if we even need to call it from the `after_save` callback at all. I think most of the time when an order is first created there's no need to tax anything...

File: app/models/spree/adjustment.rb
Comment (Matt-Yorkley) [reply app/models/spree/adjustment.rb:38]: Okay, this bit needs some explanation.

So this issue is: in Spree, adjustments are never taxable. In OFN, enterprise fee adjustments are taxable objects.

Previously when recording tax on an enterprise fee we were calculating it when the fee was created but not creating a tax adjustment. Instead we would just stick a tax amount in the `included_tax` column on the enterprise fee. This meant there was no proper link to a tax rate object and the tax couldn't be recalculated or updated. The only way to update it was to completely destroy all enterprise fees for an order and then re-create them all from scratch, via the `#recreate_all_fees!` method. It also avoids the correct (and slightly complex) logic in the `TaxRate` class and spreads the logic (and responsibility) for taxing things around the application.

Another issue was that with additional taxes (US and CA), we wouldn't even record that amount in the `included_tax` at all, and tax totals for each rate on the order were added to the order object (but with no link to which items had been taxed). So if there was a tax amount for say $5, you couldn't tell if it applied to line items, fees, shipments, or a combination.

These jumbled different methods of recording tax are OFN-specific and were added around 2014-2015 as temporary shortcuts to getting some of the tax features of Spree 2 (like having individual taxes recorded on line items), back when OFN was still on Spree 1. Those temporary shortcuts have had consequences though, and made a bit of a mess.

One of the reasons we can't go for "Option 2" in this change here is that the data basically isn't sufficient to recreate the fees and taxes. Until recently EnterpriseFee objects (the fee itself, not the adjustments it creates) were not soft-deleted. This means that lots of enterprise fee Adjustments in the database are now associated to EnterpriseFee objects that no longer exist. There's not enough information to reason about which tax rates were used when they were created or what amounts of tax are applied.

Comment (Matt-Yorkley) [reply app/models/spree/adjustment.rb:38]: So... when we create proper tax adjustments, they have a few requirements. Firstly the object being taxed must respond to `#tax_category`, which lets us figure out what rate(s) to apply. Secondly, the taxable object must respond to `#amount`, which is used in the calculation.

I've been trying to look at ways to do this, and this seems to be the simplest. So: yes, adjustments can have adjustments. In this case we could have an enterprise fee adjustment with an amount of $5 and a certain tax category. To tax it, we pass it to `Spree::TaxRate#adjust`. It'll add a tax adjustment on it which records the amount of tax, the rate used, whether it's additional or included etc, and we then have a tax adjustment associated to the fee adjustment. That association is already polymorphic (the "adjustable"). So in this case the "adjustable" object for tax recorded on a fee is: the fee (an adjustment).

Does that make sense.?

Comment (Matt-Yorkley) [reply app/models/spree/adjustment.rb:38]: I know it sounds a bit wild, but it works quite well with very minimal changes. I couldn't really see alternative options for the problem that we need to tax objects that are (fee) _adjustments_, and in order to tax things correctly we need to _add an associated tax adjustment_ to that object.

Comment (Matt-Yorkley) [reply app/models/spree/adjustment.rb:38]: This is the main bit of adapting needed to bringing the tax and adjustments changes from Spree 2.2. Everything else fits in quite well, but this needed an OFN-specific solution. It would be good to get @mkllnk 's thoughts on this as well.

File: app/controllers/spree/admin/orders_controller.rb
Comment (Matt-Yorkley) [reply app/controllers/spree/admin/orders_controller.rb:43]: :100: Yeah I think ultimately we need a new service (or two) that handles this stuff with updating orders after they've changed. There's a lot of logic around that which is currently spread out. There's a couple more changes related to the new `Spree::ItemAdjustments` class I'd like to bring in next (and then the adjustments changes will basically be finished) and then it would be really good to assess how we can improve these things.

File: app/models/spree/adjustment.rb
Comment (Matt-Yorkley) [reply app/models/spree/adjustment.rb:38]: Another consequence of _not_ recording amounts of tax in proper tax adjustments (which means we have tax amounts attached to objects, but without a reference to a tax rate) is the madness that goes on with the `TaxRateFinder` class, and if you read the code comments [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/a1317be19b76d1953e38f9e47cbc8e5304d275e6/app/services/tax_rate_finder.rb#L59-L75) you'll see why I want to delete that class as soon as possible. This change brings us closer to be able to do that :fire:

File: app/controllers/spree/admin/orders_controller.rb
Comment (sauloperez) [reply app/controllers/spree/admin/orders_controller.rb:43]: but my point is, defining some boundaries by improving the design should make this easier to change. I know it would then not be possible to apply commits Spree's into our repo but it'd help us in the long run.

File: app/models/spree/order.rb
Comment (sauloperez) [reply app/models/spree/order.rb:377]: well, that's even better!

**Comment by sauloperez** [comment]
Shall we have a design symposium some day @Matt-Yorkley . We could maybe help plan and aid in implementing the design improvements you suggest.

**Comment by Matt-Yorkley** [comment]
> Shall we have a design symposium some day @Matt-Yorkley ? We could maybe help plan and aid in implementing the design improvements you suggest.

Sure. Do you think we need a dev meeting to discuss the **taxes on fees** issue? Or do you mean broader discussion/planning for some of the improvements we've talked about around callbacks/services/abstractions?

File: app/models/spree/adjustment.rb
Comment (mkllnk) [reply app/models/spree/adjustment.rb:38]: > I couldn't really see alternative options for the problem that we need to tax objects that are (fee) adjustments

The only other idea I would have is to make enterprise fees their own object which then has adjustments. But I can't think of any benefit of this except being more explicit and having less polymorphic complexity.

Your explanation of adjustable adjustments sounds reasonable and like a good way forward. I, too, want to get rid of the TaxRateFinder which should have been named InsufficientTaxRateGuesser.

Comment (mkllnk) [reply app/models/spree/adjustment.rb:38]: Well, it finds tax rates, just not always the right ones.

Comment (Matt-Yorkley) [reply app/models/spree/adjustment.rb:38]: > The only other idea I would have is to make enterprise fees their own object which then has adjustments.

Hmm... yeah that's not a bad idea. They would have to behave almost exactly like adjustments... I think if we went down that road, the data migration we would need in order to convert all existing fee adjustments to a different kind of object (with a different database table?) might end up being totally crazy. :grimacing:

**Comment by sauloperez** [comment]
I'd scope it to this particular PR.

File: app/models/spree/adjustment.rb
Comment (sauloperez) [reply app/models/spree/adjustment.rb:38]: yes, it sounds like the other reasonable thing to do, but we could also wait and see if there's any extra value in going down that path. It seems that it would just add duplication now. The complexity of the polymorphism would still be there.

File: app/models/spree/tax_rate.rb
Comment (luisramos0) [inline app/models/spree/tax_rate.rb:60]: this class is so complicated. I think this method (adjust) could go directly to a service, by itself with some good unit testing.

**Review by luisramos0 — APPROVED** [review]
Matt, it's incredible how you can navigate this problem and code! Congrats!

I think the explanations and code looks pretty nice :ok_hand:

So, for this particular PR, more than improving the code I'd focus on reviewing in detail our test cases and then make sure they are automated.

**Comment by Matt-Yorkley** [comment]
There's a lot of missing test coverage on the behavior in the `BulkLineItemsController` that's relevant here. I'll make a separate PR for it, and we'll have to merge that before moving ahead with this one.

**Comment by sauloperez** [comment]
wise move @Matt-Yorkley :clap:

**Comment by Matt-Yorkley** [comment]
The bad news is: I started writing tests for that controller and they revealed a serious bug in master. The good news is: that bug is actually fixed by the changes in this PR :smile:

File: db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb
Comment (mkllnk) [inline db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb:88]: Is this simpler? :shrug: 
```suggestion
    tax_category&.tax_rates&.first
```

File: db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb
Comment (mkllnk) [inline db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb:42]: How long does this migration run for with production data?

**Review by mkllnk — APPROVED** [review]
What a fight! The final boss made it to round 32 and is now defeated. :wink: 

No, seriously, impressive work. :+1: 

It's time to flesh out the testing instructions. Are you happy to move it to test-ready or shall we wait for another review?

File: app/models/spree/tax_rate.rb
Comment (Matt-Yorkley) [reply app/models/spree/tax_rate.rb:84]: I think this is just a direct Spree import. It does look a bit weird, but I guess they wanted a comment on each part of the conditional to explain it carefully?

File: db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb
Comment (Matt-Yorkley) [reply db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb:88]: Sure :ok_hand:

File: db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb
Comment (Matt-Yorkley) [reply db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb:42]: I haven't checked recently, I'll rebase and try it now :+1:

**Comment by Matt-Yorkley** [comment]
> It's time to flesh out the testing instructions.

:100: I'll start writing up a plan. We should probably do both dev-tests and regular tests, and we'll want to do it with production data from "additional" tax instances and "included" tax instances, which we now have on UK Staging (Canadian data) and FR Staging (French data).

File: db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb
Comment (Matt-Yorkley) [reply db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb:42]: For instances that use **additional tax**, there's not much that changes. This is the migration with CA prod data:
```
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrating ===================
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrated (2.3167s) ==========
```
This is because there aren't really any enterprise fee adjustments that have a non-zero value for `included_tax` :point_up: 

With French data it's very different. The total number of enterprise fee adjustments with non-zero included tax amounts is: **152,981.** 
```
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrating ===================
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrated (1668.5366s) =======
```
Total time: 1668 seconds (~27 minutes). :grimacing: 

I previously looked at doing this with raw SQL, but I think there's too much complexity...

Comment (Matt-Yorkley) [reply db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb:42]: It's marginally when eager-loading both the originator and the adjustable:
```
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrating ===================
== 20210406161242 MigrateEnterpriseFeeTaxAmounts: migrated (1405.0915s) =======
```
~23 minutes.

**Comment by Matt-Yorkley** [comment]
### Things to Test

These are the bits we need to look at. Testing suggestions welcome!

**Enterprise Fee Tax Migration**

The first big thing we need to test is the migration, and it only really affects included-tax instances. There are unit tests on the migration in this PR. Running it with French Prod data and taking a look shows it created the correct number of adjustments (tax amounts recorded in the `included_tax` column for enterprise fees have been migrated to separate tax adjustment objects):
```
irb(main):001:0> Spree::Adjustment.enterprise_fee.where('included_tax <> 0').count
   (1447.2ms)  SELECT COUNT(*) FROM "spree_adjustments" WHERE "spree_adjustments"."originator_type" = $1 AND (included_tax <> 0)  [["originator_type", "EnterpriseFee"]]
=> 152981
irb(main):002:0> Spree::Adjustment.tax.where(adjustable_type: "Spree::Adjustment").count
   (344.0ms)  SELECT COUNT(*) FROM "spree_adjustments" WHERE "spree_adjustments"."originator_type" = $1 AND "spree_adjustments"."adjustable_type" = $2  [["originator_type", "Spree::TaxRate"], ["adjustable_type", "Spree::Adjustment"]]
=> 152981
```
Looking through the results, they all seem to match up and look correct. Do we want some dev-tests on this?

**Additional Taxes**

On additional-tax instances taxes on new orders should be the same as before, but should add taxes on each taxable item instead of just subtotals on the order. There's a good amount of test coverage on this.

**"Legacy" Taxes**

Old orders on additional-tax instances should still have tax totals on the order, per tax rate. If we hit the "Save and recalculate fees" button on an order, the old taxes on the order should be removed and replaced with new ones on the items. There's tests on this as well. This "upgrading" of legacy taxes should happen any time we touch an old older and update it. 

**Reports**

I guess we should have a look through reports that use tax adjustments and check they're okay?

File: app/models/spree/order.rb
Comment (luisramos0) [inline app/models/spree/order.rb:601]: can you clarify what is "use additional taxes"?
can you clarify what are "old orders"? Orders before a specific version? Orders created before this PR?

File: app/models/spree/tax_rate.rb
Comment (luisramos0) [inline app/models/spree/tax_rate.rb:35]: I think this could be extracted to a method with a good name:

`all.includes(zone: { zone_members: :zoneable }).load`

all_tax_rates_with_members?

File: app/models/spree/tax_rate.rb
Comment (luisramos0) [reply app/models/spree/tax_rate.rb:84]: comments can always be method names. treatment 2 here: https://refactoring.guru/smells/comments

```ruby

 def potentially_applicable?(order_tax_zone)
     rate_zone_matches_order_tax_zone? ||
         rate_zone_contains_order_tax_zone? ||
         rate_zone_is_default_zone?
  end

def rate_zone_matches_order_tax_zone?
      zone == order_tax_zone
end

def rate_zone_contains_order_tax_zone?
      zone.contains?(order_tax_zone)
end

def rate_zone_is_default_zone?
      included_in_price? && zone.default_tax
end

```

File: db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb
Comment (luisramos0) [inline db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb:62]: do we keep the value included_tax in the enterprise fee? Ideally, we would clear it and just have a flag to indicate the tax in the adjustment is included in the price.

File: engines/order_management/app/services/order_management/order/updater.rb
Comment (luisramos0) [inline engines/order_management/app/services/order_management/order/updater.rb:218]: I think it's better if we explain here or somewhere what are "legacy taxes"

**Review by luisramos0 — APPROVED** [review]
I had a second look. Amazing stuff.

**Comment by andrewpbrett** [comment]
I moved to Test Ready based on 2 approving reviews but will take another look through it too.

File: db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb
Comment (mkllnk) [reply db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb]: Just wondering if we need to deploy this with maintenance. But I don't think that's necessary because this wouldn't cause any pages to raise errors, right?

Comment (Matt-Yorkley) [reply db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb]: > deploy this with maintenance

:100:

Definitely! Active users could be changing objects that this migration is trying to alter. So for example new enterprise fee adjustments could be created using the "old" code which would then not be updated correctly by the migration. A 23 minute time window leaves plenty of room for creating lots of little data inconsistencies!

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

I'll start off with staging-UK - the added tax case. 

I think an approach to validate some of our expectations on this PR could be to compare the same or similar orders placed before and after these changes. Here we go...

Below we find two different cases, in which the same order was placed, for customers with billing address within state `Ontario`, shopfront and backoffice. Before staging this PR - left; after staging - right side.

Case 1) 

- Product A on Tax Category` 20 | Ontario HST | HST = 13% where relevant`
- Product B on Tax Category `13 | Saskatchewan GST + PST   | GST + PST = 11% where relevant`
- shipping method on tax category  `22 | taxable                  | taxable product          ` (which is 13% for Ontario as well)
- enterprise fee set to inherit tax from product
- a transaction fee

image

image

Reports -` /admin/reports/sales_tax`:
image
image


**Some remarks**
- The order total is the same (before and after the PR) :+1:  
- Customers billing address is Ontario so, product B is not taxed
- The enterprise fee is taxed as the product is (which is correct)
- The tax value applied over the product does not appear in the order confirmation for the customer. This is seen when observing the difference:
0.78 (tax from item price w/ fee) - 0.52 (tax from fee) = 0.26 (tax from item price). This does appear in the BO (second pic right side).
- The tax labels are incorrect - they seem to relate to the calculated tax adjustment, after this PR, for example:
`Ontario - HST 26% - $0.26`, although the tax rate is 13%.
- On the Backoffice: The tax amounts are now appearing on the **line item adjustments** section :tada: 


Case 2)
- Supplier A, Product A and Supplier B, Product B - both on Tax Category  `20 | Ontario HST           | HST = 13% where relevant `
- shipping method without tax category
- an enterprise fee (Supplier B), same tax category
- Supplier B is set not to charge taxes

image

image

Reports -` /admin/reports/sales_tax`:
image
image

**Some remarks**

Pretty much reproduces the scenario case 1) above. This was done to check the scenario described here; the conclusion is, (before and after this PR) the distributor settings override the supplier settings.

**To summarize so far:**
- the tax labels are incorrect
- shopfront does not display line item tax adjustments; it shows taxes on enterprise fees
- backoffice is the other way around: displays line item adjustments but it does not display tax adjustments on enterprise fees

This looks pretty good I think?  It's challenging to track all the differences, but I think its mostly working as expected. What do you think @Matt-Yorkley ?

Now to check staging-FR...

**Comment by Matt-Yorkley** [comment]
Awesome. Yeah, that looks great. :+1:

FYI if you try _really_ hard there's a small chance you might hit a rounding issue where the old and new versions are different by `0.01` due to the way maths works in that; if you calculate lots of small subtotals (one for each item) that are _individually_ rounded and then add them together you can end up with a slightly different number than if you calculate a single grand total all at once and then apply rounding once (which is what used to happen with additional tax). But I see you haven't encountered that. Also there's nothing we could reasonably do about it anyway :sweat_smile:

**Comment by filipefurtad0** [comment]
Ok, I checked staging-AU instead, which had a GST tax rates set for New Zealand:

- NZ | GST | 20%
- NZ | Shipping - NZ | 10%


As before, orders placed before and after this PR show a disagreement between the label and the actually applied tax. But other than that, there are no big changes in this case, orders match and tax rates are applied correctly:

image

image


Reports work as before too - `/admin/reports/xero_invoices`:
image

I just noticed I was looking at the wrong report on the previous testing notes (adding a quick update on that).

I think this is good to go, but perhaps we'll find time for a round of testing between amongst instance managers before the release?

**Comment by Matt-Yorkley** [comment]
Hmm yeah some of the text in those labels is definitely off, I'll take a look at that. Pretty sure it's a UI display issue as opposed to an issue with the actual amounts being applied :+1:

**Comment by filipefurtad0** [comment]
Totally agree @Matt-Yorkley - it seems like a UI thing; the values are correctly applied as far as I could gather.

**Comment by Matt-Yorkley** [comment]
I just pushed a commit which should fix that label issue :+1:

**Comment by filipefurtad0** [comment]
Looks great now :muscle: 

Staging-UK - added tax:
image

Staging-AU - included tax:
image

The labels are correct :+1:

File: db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb
Comment (Matt-Yorkley) [reply db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb]: Update: I just deployed this on `fr-staging` which currently has a very recent production dataset and it only took **~17 minutes**, which is a lot better. Maybe the ofn-install task that clears old data before migrating made a difference?

**Pull request merged**
